contributors/design-proposals/encryption.md

```diff
@@ -0,0 +1,429 @@
+    Slot int
+}
+```
+
+```go
+type KeyEncryptionKeyDatabase struct {
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+    Keys []string
+}
+```
+
+To enable encryption a user must first create a KEK DB file and tell the API
+server to use it with `--encryption-provider-config=/path/to/config`.  The
+file will be a simple YAML file that lists all of the keys:
+
+```yaml
+kind: KeyEncryptionKeyDatabase
+version: v1
+keys:
+ - foo
+ - bar
+ - baz
+```
+
+Then a user calls PUT on a /rotate API endpoint the first time:
+
+1. A new encryption key (unencrypted DEK) is created.
+1. Encrypt DEK with KEK[1]
+1. The list of DEKs being used by the master is updated in memory so that the
+   new key is in the list of read keys.
+1. The list of DEKs being used by the master is updated in etcd so that the
+   new key is in the list of read keys available to all masters.
+1. Confirm that all masters have the new DEK for reading. Key point here is that
+   all readers have the new key before anyone writes with it.
+1. The list of DEKs being used by the master is updated in memory so that the
+   new key is is the current write key.
+1. The list of DEKs being used by the master is updated in etcd so that the new
+   key is the current write key and is available to all masters. It doesn't
+   matter if there's some masters using the new key and some using the old key,
+   since we know all masters can read with the new key. Eventually all masters
+   will be writing with the new key.
+1. All secrets are re-encrypted with the new key.
+
+After N rotation calls:
+
+1. A new encryption key (unencrypted DEK) is created.
+1. Encrypt DEK with KEK[N+1]
+
+Each rotation generates a new KEK and DEK. Two DEKs will be in-use temporarily
+during rotation, but only one at steady-state.
+
+Pros:
+
+ - Most closely matches the pattern that will be used for integrating with
+   external encryption systems. Amazon KMS, Google KMS and HSM would eventually
+   serve the purpose of KEK storage rather than local disk.
+
+Cons:
+
+ - End state is still KEKs on disk on the master. This is equivalent to the much
+   simpler list of keys on disk in terms of key management and security.
+   Complexity is much higher.
+ - Coordination of keys between a deployment with multiple masters will require
+   shutting down all masters, and manually generating and providing a key in the
+   key file. Same as keys on disk.
+
+### Option 4: Protocol for KEK agreement between masters
+
+TODO: write a proposal for coordinating KEK agreement among multiple masters and
+having the KEK be either user supplied or backed by external store.
+
+
+## External providers
+
+It should be easy for the user to substitute a default encryption provider for
+one of the following:
+
+* A local HSM implementation that retrieves the keys from the secure enclave
+  prior to reusing the AES-GCM implementation (initialization of keys only)
```

Comment:
I think? this is saying there's only one layer of keys, and they are retrieved from the HSM. So it's like option 1 with the keys in the HSM instead of on disk. @philips wrote this originally I believe.